lib/sycamore/sycamore/query/schema.py

```diff
@@ -90,16 +90,30 @@ def get_schema(self) -> OpenSearchSchema:
                             sample_type = type(sample_value)
                         else:
                             t = type(sample_value)
-                            if str(t) != str(sample_type):
-                                if str(t) == "<class 'int'>" and str(sample_type) == "<class 'float'>":
-                                    pass  # compatible
-                                elif str(sample_type) == "<class 'int'>" and str(t) == "<class 'float'>":
-                                    sample_type = t  # upgrade
-                                else:
-                                    logger.warning(
-                                        "Got multiple sample types for key"
-                                        + f" {key}: {sample_type} and {t}. Keeping the former"
-                                    )
+
+                            # Need to check for compatibility between the sample type and the previously
+                            # seen type, and possibly upgrade the sample type.
+                            if sample_type == int and t == float:
+                                # Upgrade from int to float.
+                                sample_type = t
+                            elif sample_type == float and t == int:
+                                # No need to change.
+                                pass
+                            elif sample_type == list and t != list:
+                                # Upgrade from singleton to list.
+                                sample_value = [sample_value]
+                            elif sample_type != list and t == list:
+                                # Need to upgrade our sample type to match the new list type.
+                                sample_type = t
+                                samples = {str([x]) for x in samples}
+                            elif sample_type != t:
```

Comment:
Oh, you mean just to special case when sample_type is already str?

Reply:
Yes, I want to not get lots of log messages since we're auto handling this case.
Maybe you add a
```
if not reported[key]:
  logger.warning("Auto coercing <type> to str")
  reported[key] = True
```
to make sure that we log in the case where the str example was found first would be better (and then set reported[key] = True in the other branch)

Reply:
Good point, done.